include/pybind11/common.h

```diff
@@ -321,6 +321,7 @@ struct internals {
     std::unordered_map<const void *, void*> registered_types_py;       // PyTypeObject* -> type_info
     std::unordered_multimap<const void *, void*> registered_instances; // void * -> PyObject*
     std::unordered_set<std::pair<const PyObject *, const char *>, overload_hash> inactive_overload_cache;
+    std::unordered_map<std::type_index, std::vector<bool (*)(PyObject *, void *&)>> direct_conversions;
```

Comment:
For instance, the generic_caster won't be able to deal with a dtype that has not have a proper ``py::detail::type_info`` record, so it seems unnecessary to track the associated conversions with a new global data structure.